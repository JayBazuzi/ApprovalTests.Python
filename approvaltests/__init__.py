"""Approvaltests module."""
from .version import *
from .utilities import *
from .approvals import *
from .asserts import *

__author__ = "Chris Lucian; Llewellyn Falco; Jim Counts; Tim Preece"
__version__ = version_number
