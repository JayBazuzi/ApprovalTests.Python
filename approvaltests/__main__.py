from .verify import verify_using_commandline_arguments

verify_using_commandline_arguments()