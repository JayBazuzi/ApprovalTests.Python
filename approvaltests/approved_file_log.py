from pathlib import Path

APPROVAL_TESTS_TEMP_DIRECTORY: str = ".approval_tests_temp"


class ApprovedFilesLog:
    @staticmethod
    def clear_log_file() -> None:
        ApprovedFilesLog.get_approved_files_log().write_text("")

    @staticmethod
    def get_approved_files_log() -> Path:
        path = Path("%s/.approved_files.log" % APPROVAL_TESTS_TEMP_DIRECTORY)
        path.parent.mkdir(parents=True, exist_ok=True)

        return path

    @staticmethod
    def log(approved_file):
        with ApprovedFilesLog.get_approved_files_log().open(mode="a") as file:
            file.write(f"{approved_file}\n")


ApprovedFilesLog.clear_log_file()
