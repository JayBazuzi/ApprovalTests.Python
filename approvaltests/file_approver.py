import filecmp
import os
import pathlib
from typing import Optional, Callable

from approvaltests.core.comparator import Comparator
from approvaltests.core.namer import Namer
from approvaltests.core.reporter import Reporter
from approvaltests.core.writer import Writer


def exists(path: str) -> bool:
    return os.path.isfile(path)


class ReporterNotWorkingException(Exception):
    def __init__(self, reporter: Reporter):
        super().__init__(f"Reporter {reporter} failed to work!")


class FileComparator(Comparator):
    def compare(self, received_path: str, approved_path: str) -> bool:
        if not exists(approved_path) or not exists(received_path):
            return False
        if filecmp.cmp(approved_path, received_path):
            return True
        try:
            approved_raw = pathlib.Path(approved_path).read_text()
            approved_text = approved_raw.replace("\r\n", "\n")
            received_raw = pathlib.Path(received_path).read_text()
            received_text = received_raw.replace("\r\n", "\n")
            return approved_text == received_text
        except BaseException:
            return False


class FileApprover:
    previous_approved = []
    allowed_duplicates = []

    @staticmethod
    def verify(
        namer: Namer, writer: Writer, reporter: Reporter, comparator: Comparator
    ) -> Optional[str]:
        approved = namer.get_approved_filename()
        received = namer.get_received_filename()

        if (FileApprover.is_this_a_multiple_verify(approved) ):
            return (
                f"We noticed that you called verify more than once in the same test. Is that what you want to do?\n"
                f"\tApproved file name is: {approved}\n"
            )
        FileApprover.previous_approved.append(approved)

        # The writer has the ability to change the name of the received file
        received = writer.write_received_file(received)
        verified = FileApprover.verify_files(approved, received, reporter, comparator)

        if not verified:
            return (
                f"Approval Mismatch, received != approved\n"
                f"\tApproved: {approved}\n"
                f"\tReceived: {received} "
            )
        return None

    @staticmethod
    def is_this_a_multiple_verify(approved):
        return approved in FileApprover.previous_approved \
            and not FileApprover.is_duplicate_allowed(approved)

    @staticmethod
    def is_duplicate_allowed(approved):
        for allowed in FileApprover.allowed_duplicates:
            if allowed(approved):
                return True
        return False

    @staticmethod
    def verify_files(
        approved_file: str,
        received_file: str,
        reporter: Reporter,
        comparator: Comparator,
    ) -> bool:
        if comparator.compare(received_file, approved_file):
            os.remove(received_file)
            return True

        worked = reporter.report(received_file, approved_file)
        if not worked:
            raise ReporterNotWorkingException(reporter)
        return False

    @staticmethod
    def add_allowed_duplicates(is_duplicate_allowed: Callable[[str],bool]):
        FileApprover.allowed_duplicates.append(is_duplicate_allowed)
