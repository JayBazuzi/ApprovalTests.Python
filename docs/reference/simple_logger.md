# Simple Logger

<!-- toc -->
## Contents

  * [How tos](#how-tos)<!-- endToc -->

## How tos
* [How log method parameter values](../how_to/log_method_parameter_values.md)
