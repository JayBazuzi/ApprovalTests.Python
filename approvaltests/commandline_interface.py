from pathlib import Path

from approvaltests import verify
from build.lib.approvaltests.commandline_interface import parse_arguments
from namer.cli_namer import CliNamer


# simplify the following lines of code from commandline_interface.py

# 1.  Modify so our future-selves would recognize it
# >>> 2.  Make a test
# 3.  Refactor parse_arguments to be part of a class (suggested by Jay)

# NEw code...
# received = args.received or stdin.read()
# Old code...
# received = args.received
# if args.received is None:
#    received = stdin.read()
# return (args.id, received)



def verify_using_commandline_arguments():
    test_id, received = parse_arguments()
    verify_with_id(received, test_id)


def verify_with_id(received, test_id):
    verify(received, namer=CliNamer(test_id=test_id))
    print(f"Test Passed: {test_id}")


if __name__ == "__main__":
    verify_using_commandline_arguments()


def list_approved_files_in_test_directory(test_dir: Path):
    return [file.name for file in test_dir.iterdir() if file.name.endswith(".approved.txt")]
