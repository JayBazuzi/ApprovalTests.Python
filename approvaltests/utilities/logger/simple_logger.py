import inspect
from contextlib import contextmanager
from typing import Iterator

from approvaltests.utilities.string_wrapper import StringWrapper
from approvaltests.namer import StackFrameNamer


class SimpleLogger:
    logger = print
    tabbing = 0

    @staticmethod
    def log_to_string():
        buffer = StringWrapper()
        SimpleLogger.logger = buffer.append
        return buffer

    @staticmethod
    @contextmanager
    def use_markers() -> Iterator[None]:
        stack = inspect.stack(1)[2]
        method_name = stack[3]
        filename = StackFrameNamer.get_class_name_for_frame(stack)
        expected = f"-> in: {method_name}(){filename}"
        SimpleLogger.log(expected)
        SimpleLogger.tabbing = SimpleLogger.tabbing + 1
        yield
        SimpleLogger.tabbing = SimpleLogger.tabbing - 1
        expected = f"<- out: {method_name}(){filename}"
        SimpleLogger.log(expected)
        pass

    @staticmethod
    def log(log_output):
        tabbing = '  ' * SimpleLogger.tabbing
        SimpleLogger.logger(f"{tabbing}{log_output}\n")
