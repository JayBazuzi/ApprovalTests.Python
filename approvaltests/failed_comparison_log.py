from pathlib import Path

from approvaltests.approved_file_log import APPROVAL_TESTS_TEMP_DIRECTORY


class FailedComparisonLog:
    @staticmethod
    def clear_log_file() -> None:
        FailedComparisonLog.get_failed_comparison_log().write_text("")

    @staticmethod
    def get_failed_comparison_log() -> Path:
        path = Path("%s/.failed_comparison.log" % APPROVAL_TESTS_TEMP_DIRECTORY)
        path.parent.mkdir(parents=True, exist_ok=True)

        return path

    @staticmethod
    def log(received_file, approved_file):
        with FailedComparisonLog.get_failed_comparison_log().open(mode="a") as file:
            file.write(f"{received_file} -> {approved_file}\n")


FailedComparisonLog.clear_log_file()
