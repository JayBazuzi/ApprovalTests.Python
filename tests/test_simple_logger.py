import datetime
import os
from pathlib import Path

from approvaltests import verify, Options
from approvaltests.utilities.logger.simple_logger import SimpleLogger


def test_warnings():
    def scrubber(text: str) -> str:
        return text.replace(__file__, "test_simple_logger.py")

    output = SimpleLogger.log_to_string()
    SimpleLogger.logger.log_stack_traces = True
    text = "EVERYTHING IS AWFUL!!!!!!"
    try:
        raise Exception("EVERYTHING IS exceptionally AWFUL!!!!!!")
    except Exception as e:
        exception = e
    SimpleLogger.warning(text)
    SimpleLogger.warning(exception)
    SimpleLogger.warning(text, exception)
    verify(output, options=Options().with_scrubber(scrubber))


def log_from_inner_method():
    with SimpleLogger.use_markers() as m:
        name = "Example"
        SimpleLogger.variable("name", name)
        for _ in range(0, 142):
            SimpleLogger.hour_glass()


def test_standard_logger():
    output = SimpleLogger.log_to_string()
    with SimpleLogger.use_markers() as m:
        log_from_inner_method()

    verify(output)


def test_timestamps():
    output = SimpleLogger.log_to_string()
    count = -1

    def create_applesauce_timer():
        dates = [
            datetime.datetime.fromtimestamp(0.0),
            datetime.datetime.fromtimestamp(0.5),
            datetime.datetime.fromtimestamp(2.0),
            datetime.datetime.fromtimestamp(1050),
            datetime.datetime.fromtimestamp(1052),
        ]
        nonlocal count
        count = count + 1
        return dates[count]

    SimpleLogger.logger.timer = create_applesauce_timer
    SimpleLogger.show_timestamps(True)
    SimpleLogger.event("1")
    SimpleLogger.event("2")
    SimpleLogger.event("3")
    SimpleLogger.event("4")
    SimpleLogger.warning(exception=Exception("Oh no you didn't!"))
    verify(output)


def test_variable():
    names = ["Jacqueline", "Llewellyn"]
    output = SimpleLogger.log_to_string()
    SimpleLogger.variable("names", names)
    verify(output)


def verify_toggle(toggle_name, toggle):
    SimpleLogger.show_all(True)
    SimpleLogger.event(f"Toggle Off {toggle_name}")
    toggle(False)
    log_everything()


def test_switching() -> None:
    output = SimpleLogger.log_to_string()

    verify_toggle("None", lambda a: SimpleLogger.show_all(True)),
    verify_toggle("All", SimpleLogger.show_all),
    verify_toggle("Query", SimpleLogger.show_queries),
    verify_toggle("Message", SimpleLogger.show_messages),
    verify_toggle("Variable", SimpleLogger.show_variables),
    verify_toggle("Hour Glass", SimpleLogger.show_hour_glass),
    verify_toggle("Markers", SimpleLogger.show_markers),
    verify_toggle("Events", SimpleLogger.show_events),

    verify(output)


def log_everything() -> None:
    with SimpleLogger.use_markers() as m:
        SimpleLogger.query("Select * from people")
        SimpleLogger.variable("Nonsense", "foo")
        SimpleLogger.event("Testing")
        SimpleLogger.message("Something random")
        for a in range(1, 13):
            SimpleLogger.hour_glass()
        try:
            infinity = 1 / 0
        except Exception as e:
            SimpleLogger.warning(exception=e)
