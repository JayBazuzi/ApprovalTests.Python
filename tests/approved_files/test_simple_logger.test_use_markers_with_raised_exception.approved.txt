-> in: throw_exception()test_simple_logger
<- out: throw_exception()test_simple_logger
*******************************************************************************************
Exception: Everything is awflu!!?
*******************************************************************************************
